dataless-classifier/src/main/java/edu/illinois/cs/cogcomp/datalessclassification/representation/esa/MemoryBasedESA.java

```diff
@@ -0,0 +1,392 @@
+import edu.illinois.cs.cogcomp.datalessclassification.representation.AEmbedding;
+import edu.illinois.cs.cogcomp.datalessclassification.util.SparseVector;
+import edu.illinois.cs.cogcomp.datalessclassification.util.SparseVectorOperations;
+import org.cogcomp.Datastore;
+import org.cogcomp.DatastoreException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Computes ESA Embedding for a query. Loads up all the required DataStructures in memory, and is
+ * thus quite fast.
+ * 
+ * @author yqsong@illinois.edu
+ * @author shashank
+ */
+
+public class MemoryBasedESA extends AEmbedding<Integer> {
+    private static Logger logger = LoggerFactory.getLogger(MemoryBasedESA.class);
+
+    private static Map<String, SparseVector<Integer>> vectors;
+    private static Map<String, Double> wordIDF;
+    private static Map<Integer, String> pageIdTitleMapping;
+
+    private int dimensions;
+
+    private enum FileType {
+        Embedding, Mapping
+    }
+
+    public MemoryBasedESA() {
+        this(new ESADatalessConfigurator().getDefaultConfig());
+    }
+
+    public MemoryBasedESA(ResourceManager config) {
+        this(config.getInt(ESADatalessConfigurator.ESA_DIM));
+    }
+
+    public MemoryBasedESA(int embSize) {
+        dimensions = embSize;
+    }
+
+    private void loadVectors() {
+        if (vectors == null) {
+            try {
+                File inputFile = getFile(FileType.Embedding);
+                logger.info("Reading ESA Embeddings from " + inputFile.getAbsolutePath());
+
+                vectors = new HashMap<>();
+                wordIDF = new HashMap<>();
+
+                int count = 0;
+
+                BufferedReader reader = new BufferedReader(new FileReader(inputFile));
+
+                String line;
+
+                while ((line = reader.readLine()) != null) {
+                    line = line.trim();
+
+                    if (line.length() > 0) {
+                        String[] arr = line.split("\t");
+
+                        String word = arr[0];
+
+                        double idf = Double.parseDouble(arr[1]);
+                        wordIDF.put(word, idf);
+
+                        String[] conceptValues = arr[2].split(";");
+
+                        Map<Integer, Double> map = new HashMap<>();
+
+                        for (int i = 0; i < conceptValues.length; i++) {
+                            String[] tokens = conceptValues[i].split(",");
+                            map.put(Integer.parseInt(tokens[0]), Double.parseDouble(tokens[1]));
+                        }
+
+                        SparseVector<Integer> sparseVector = new SparseVector<>(map);
+                        vectors.put(word, sparseVector);
+                    }
+
+                    count++;
+
+                    if (count % 100000 == 0)
+                        logger.info("#ESA embeddings read: " + count);
+                }
+
+                reader.close();
+                logger.info("Done.");
```

Comment:
either close the reader in a 'finally' block, or use try-with-resources

Reply:
Done.